north/src/runtime/keys.rs

```diff
@@ -36,40 +32,40 @@ pub enum Error {
     let mut signing_keys = HashMap::new();
-    let mut key_dir = fs::read_dir(&key_dir).await.map_err(|e| Error::Io {
-        context: format!("Failed to open {}", key_dir.display()),
-        error: e,
-    })?;
+    let mut key_dir = fs::read_dir(&key_dir)
+        .await
+        .context(format!("Failed to open {}", key_dir.display()))?;
     while let Some(entry) = key_dir.next().await {
-        let entry = entry.map_err(|e| Error::Io {
-            context: "Invalid key dir entry".to_string(),
-            error: e,
-        })?;
-        // .context("Invalid key dir entry")?;
-        let path = entry.path();
-        if let Some(extension) = path.extension() {
-            if extension == "pub" && path.is_file() {
-                if let Some(key_id) = path.file_stem().map(|s| s.to_string_lossy().to_string()) {
-                    info!("Loading signing key {}", key_id);
-                    let mut sign_key_file = fs::File::open(&path).await.map_err(|e| Error::Io {
-                        context: format!("Failed to open {}", path.display()),
-                        error: e,
-                    })?;
-                    let mut key_bytes = Vec::new();
-                    sign_key_file
-                        .read_to_end(&mut key_bytes)
-                        .await
-                        .map_err(|e| Error::Io {
-                            context: format!("Failed to read {}", path.display()),
-                            error: e,
-                        })?;
-                    let key = PublicKey::from_bytes(&key_bytes).map_err(|_| {
-                        Error::Signature(format!("Signature error for key from {}", path.display()))
-                    })?;
-                    signing_keys.insert(key_id, key);
-                }
-            }
+        let path = entry.context("Invalid key dir entry".to_string())?.path();
+        if path.extension().filter(|ext| *ext == "pub").is_none() || !path.is_file() {
+            continue;
+        }
+
+        if let Some(key_id) = path.file_stem().map(|s| s.to_string_lossy().to_string()) {
+            info!("Loading signing key {}", key_id);
+            let key_bytes = fs::read(&path)
+                .await
+                .context(format!("Failed to read {}", path.display()))?;
+            let key = PublicKey::from_bytes(&key_bytes)
+                .context(format!("Signature error for key from {}", path.display()))?;
+            signing_keys.insert(key_id, key);
         }
     }
     Ok(signing_keys)
 }
+
+trait MapError<T> {
+    fn context(self, context: String) -> Result<T, Error>;
+}
+
+impl<T> MapError<T> for io::Result<T> {
+    fn context(self, context: String) -> Result<T, Error> {
+        self.map_err(|error| Error::Io { context, error })
+    }
+}
+
+impl<T> MapError<T> for Result<T, ed25519_dalek::SignatureError> {
+    fn context(self, context: String) -> Result<T, Error> {
+        self.map_err(|_| Error::Signature(context))
+    }
+}
```

Comment:
Not sure if we should declare such a general util thing in here. Why not declaring a error variant that uses `#from(io::Error)` and a "context" string e.g

```diff
diff --git a/north/src/runtime/keys.rs b/north/src/runtime/keys.rs
index a8b1378..1f61607 100644
--- a/north/src/runtime/keys.rs
+++ b/north/src/runtime/keys.rs
@@ -20,12 +20,8 @@ use tokio::{fs, io, stream::StreamExt};
 
 #[derive(Error, Debug)]
 pub enum Error {
-    #[error("Io error: {context}")]
-    Io {
-        context: String,
-        #[source]
-        error: io::Error,
-    },
+    #[error("IO error: {0}")]
+    Io(&'static str, #[source] io::Error),
     #[error("Invalid key signature: {0}")]
     Signature(String),
 }
@@ -34,9 +30,9 @@ pub async fn load(key_dir: &Path) -> Result<HashMap<String, PublicKey>, Error> {
     let mut signing_keys = HashMap::new();
     let mut key_dir = fs::read_dir(&key_dir)
         .await
-        .context(format!("Failed to open {}", key_dir.display()))?;
+        .map_err(|e| Error::Io("Failed to read key dir", e))?;
     while let Some(entry) = key_dir.next().await {
-        let path = entry.context("Invalid key dir entry".to_string())?.path();
+        let path = entry.map_err(|e| Error::Io("Failed to read key dir entry", e))?.path();
         if path.extension().filter(|ext| *ext == "pub").is_none() || !path.is_file() {
             continue;
         }
@@ -45,27 +41,11 @@ pub async fn load(key_dir: &Path) -> Result<HashMap<String, PublicKey>, Error> {
             info!("Loading signing key {}", key_id);
             let key_bytes = fs::read(&path)
                 .await
-                .context(format!("Failed to read {}", path.display()))?;
+                .map_err(|e| Error::Io("Failed to read key", e))?;
             let key = PublicKey::from_bytes(&key_bytes)
-                .context(format!("Signature error for key from {}", path.display()))?;
+                .map_err(|e| Error::Io("Failed to read pub key", e))?;
             signing_keys.insert(key_id, key);
         }
     }
```
This is less code with the same amount of information and feels more natural.

Reply:
you cheated a bit ;) you dropped the `format!`s with the context info, that would bump up the code a bit. I will change it as you show but we should look for some way to simplify all the `map_err`s everywhere, they are a bit distracting.